src/client/api/metrics.c

```diff
@@ -0,0 +1,914 @@
+		entry->update_stat.st_min = ULONG_MAX;
+		memset(&entry->fetch_stat, 0, sizeof(entry->fetch_stat));
+		entry->fetch_stat.st_min = ULONG_MAX;
+		memset(&entry->ids, 0, sizeof(entry->ids));
+		memset(&entry->idp, 0, sizeof(entry->idp));
+	}
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+}
+
+static inline void
+dc_metrics_aggr_stats(daos_metrics_stat_t *dest, daos_metrics_stat_t *src)
+{
+	dest->st_value += src->st_value;
+	if (dest->st_min > src->st_min) {
+		dest->st_min = src->st_min;
+	}
+	if (dest->st_max < src->st_max) {
+		dest->st_max = src->st_max;
+	}
+	dest->st_sum += src->st_sum;
+	dest->st_sum_of_squares += src->st_sum_of_squares;
+}
+
+static void
+dc_metrics_tls_aggr(dc_metrics_tls_data_t *value)
+{
+	int i;
+
+	/** Update stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.update_stat, \
+			&value->update_stat);
+
+	/** Fetch stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.fetch_stat, \
+			&value->fetch_stat);
+
+	/** io distribution by size */
+	metrics_tls_agg.ids.ids_updatesz += value->ids.ids_updatesz;
+	metrics_tls_agg.ids.ids_fetchsz += value->ids.ids_fetchsz;
+	for (i=0; i< DAOS_METRICS_IO_BYSIZE_COUNT; i++) {
+		metrics_tls_agg.ids.ids_updatecnt_bkt[i] += value->ids.ids_updatecnt_bkt[i];
+		metrics_tls_agg.ids.ids_fetchcnt_bkt[i]  += value->ids.ids_fetchcnt_bkt[i];
+	}
+
+	/** io distribution by protection type */
+	metrics_tls_agg.idp.idp_updatesz += value->idp.idp_updatesz;
+	for (i=0; i< DAOS_METRICS_IO_BYPTYPE_COUNT; i++) {
+		metrics_tls_agg.idp.idp_updatecnt_bkt[i] += value->idp.idp_updatecnt_bkt[i];
+		metrics_tls_agg.idp.idp_updatesz_bkt[i]  += value->idp.idp_updatesz_bkt[i];
+	}
+}
+
+static void
+dc_metrics_tls_cleanup(void *value)
+{
+	if (metrics_tls == NULL) {
+		goto out;
+	}
+	D_MUTEX_LOCK(&metrics_tls_lock);
+	d_list_del(&metrics_tls->list);
+	dc_metrics_tls_aggr((dc_metrics_tls_data_t *)value);
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+	free(metrics_tls);
+	metrics_tls = NULL;
+out:
+	return;
+}
+
+
+int
+dc_metrics_init()
+{
+	char *env;
+	int rc, i;
+
+	env = getenv("DAOS_METRICS");
+#if 0 /* Revisit */
+	is_metrics_enabled = ((env != NULL) && (strncmp(env, "ON", strlen("ON")) == 0));
+#else
+	if ((env != NULL) && (strncmp(env, "OFF", strlen("OFF")) == 0))
+		is_metrics_enabled = 0;
+	else
+		is_metrics_enabled = 1;
+#endif
+
+	if (is_metrics_enabled == 0)
+		return 0;
+
+	rc = dc_pool_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for pool, rc = %d\n", rc);
+		D_GOTO(out, rc);
+	}
+	rc = dc_cont_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for container, rc = %d\n", rc);
+		D_GOTO(out_cont, rc);
+	}
+	rc = dc_obj_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for object, rc = %d\n", rc);
+		D_GOTO(out_obj, rc);
+	}
+	rc = dc_metrics_tls_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics TLS, rc = %d\n", rc);
+		D_GOTO(out_tls, rc);
+	}
+
+	for(i=0; i<RPDIST_ENUM_IDX_CNT; i++) {
```

Comment:
(style)  spaces required around that '=' (ctx:VxV)
(style)  spaces required around that '<' (ctx:VxV)
(style)  space required before the open parenthesis '('